source/common/io/io_uring_impl.cc

```diff
@@ -0,0 +1,158 @@
+#include "source/common/io/io_uring_impl.h"
+
+#include <sys/eventfd.h>
+
+#include "envoy/extensions/io/io_uring/v3/io_uring.pb.h"
+#include "envoy/extensions/io/io_uring/v3/io_uring.pb.validate.h"
+
+#include "source/common/common/utility.h"
+
+namespace Envoy {
+namespace Io {
+
+namespace {
+
+constexpr uint32_t DefaultIoUringSize = 300;
+
+} // namespace
+
+IoUring& IoUringFactoryImpl::getOrCreate() const { return tls_->getTyped<IoUringImpl>(); }
+
+Server::BootstrapExtensionPtr
+IoUringFactoryImpl::createBootstrapExtension(const Protobuf::Message& config,
```

Comment:
This is a bit strange, this is both in core but also an extension?

Reply:
Uh, right. Originally it was not an extension. It was turned into an extension in a later commit to avoid `static thread_local` and to use the TLS API instead as suggested.

Now I'm confused. If I keep the code to be an extension io_uring won't be usable in core. Otherwise the TLS API seems to be unavailable. I'd rather drop TLS and bootstrap extension as it was originally.

Reply:
Now _I'm_ confused, so you're not the only one :) Why can't we use TLS in extensions? Why does this need to be used directly by core?

Reply:
> Why does this need to be used directly by core?

[This comment](https://github.com/envoyproxy/envoy/pull/19082#issuecomment-981977371) is the main reason.

> Why can't we use TLS in extensions?

That was about (un)availability of TLS in core, not extensions. This concern is moot probably. Still I'm not sure why using TLS is more beneficial than `static thread_local` in this case, because I bumped into [other problems with integration testing](https://github.com/envoyproxy/envoy/pull/19339#discussion_r802882948).

Reply:
OK, then let's make it core. We can use TLS from non-extensions, it's used all over core, e.g. see https://github.com/envoyproxy/envoy/blob/1f8da8e0fffed1c5b387ca2a39e6f5380bc1613f/source/common/upstream/cluster_manager_impl.h#L726.

I think one of the nice things about using Envoy TLS is you get TLS lifecycle management, which should in theory actually work better for testing etc.

Reply:
Thanks!

So far my problem with Envoy TLS in testing has been that the thread controlling integration tests doesn't initialize its TLS slot hence lacks io_uring which is needed to open a client connection. I'll experiment with enabling TLS in the testing thread.

The bootstrap extension is dropped now.